File: app/src/main/java/ru/eva/oriokslive/adapters/SchedulerFragmentAdapter.java
package ru.eva.oriokslive.adapters;

import android.os.Build;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.chauthai.swipereveallayout.SwipeRevealLayout;
import com.chauthai.swipereveallayout.ViewBinderHelper;

import java.util.List;

import ru.eva.oriokslive.R;
import ru.eva.oriokslive.interfaces.OnGroupDeleteButtonClickListener;
import ru.eva.oriokslive.interfaces.OnGroupItemClickListener;

public class SchedulerFragmentAdapter extends RecyclerView.Adapter<SchedulerFragmentAdapter.ViewHolder> {

    private List<String> groupList;
    private OnGroupItemClickListener onGroupItemClickListener;
    private OnGroupDeleteButtonClickListener onGroupDeleteButtonClickListener;
    private final ViewBinderHelper binderHelper = new ViewBinderHelper();


    public SchedulerFragmentAdapter(List<String> groupList) {
        this.groupList = groupList;
    }


    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.group_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        String group = groupList.get(position);
        holder.group.setText(group);
        binderHelper.bind(holder.swipeLayout, group);
    }

    @Override
    public int getItemCount() {
        return groupList.size();
    }

    public void addItems(List<String> groupList) {
        this.groupList = groupList;
        notifyDataSetChanged();
    }

    public void removeItem(int position) {
        groupList.remove(position);
        notifyItemRemoved(position);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final TextView group, delete, add;
        private final SwipeRevealLayout swipeLayout;
        private final LinearLayout frontLayout;

        ViewHolder(View itemView) {
            super(itemView);
            swipeLayout = itemView.findViewById(R.id.swipe_layout);
            delete = itemView.findViewById(R.id.delete);
            group = itemView.findViewById(R.id.group);
            frontLayout = itemView.findViewById(R.id.front_layout);
            add = itemView.findViewById(R.id.add);
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
                add.setVisibility(View.GONE);
            }
            frontLayout.setOnClickListener(v -> onGroupItemClickListener.onClick(v, groupList.get(getAdapterPosition())));
            add.setOnClickListener(v -> {
                onGroupItemClickListener.onClick(v, groupList.get(getAdapterPosition()));
                swipeLayout.close(true);
            });
            delete.setOnClickListener(v ->  {
                onGroupDeleteButtonClickListener.onClick(groupList.get(getAdapterPosition()), getAdapterPosition());
                swipeLayout.close(true);
            });
        }
    }

    public void setOnGroupItemClickListener(OnGroupItemClickListener onGroupItemClickListener) {
        this.onGroupItemClickListener = onGroupItemClickListener;
    }

    public void setOnGroupDeleteButtonClickListener(OnGroupDeleteButtonClickListener onGroupDeleteButtonClickListener) {
        this.onGroupDeleteButtonClickListener = onGroupDeleteButtonClickListener;
    }
}

File: app/src/main/java/ru/eva/oriokslive/ui/adapter/SchedulerFragmentAdapter.kt
package ru.eva.oriokslive.ui.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.chauthai.swipereveallayout.ViewBinderHelper
import ru.eva.oriokslive.databinding.ListItemGroupBinding


class SchedulerFragmentAdapter(
    private val itemClickListener: (String) -> Unit,
    private val addListener: (String) -> Unit,
    private val deleteListener: (String, Int) -> Unit,
) : RecyclerView.Adapter<ScheduleFragmentViewHolder>() {

    private var groups: MutableList<String> = mutableListOf()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ScheduleFragmentViewHolder(
        ListItemGroupBinding.inflate(LayoutInflater.from(parent.context), parent, false)
    )

    override fun onBindViewHolder(holder: ScheduleFragmentViewHolder, position: Int) {
        holder.bind(groups[position], itemClickListener, addListener, deleteListener)
    }

    override fun getItemCount(): Int = groups.size

    fun addItems(items: List<String>) {
        groups = items.toMutableList()
        notifyItemRangeChanged(0, groups.size)
    }

    fun removeItem(position: Int) {
        groups.removeAt(position)
        notifyItemRemoved(position)
    }
}

class ScheduleFragmentViewHolder(private val binding: ListItemGroupBinding) :
    RecyclerView.ViewHolder(binding.root) {

    private val binderHelper = ViewBinderHelper()

    fun bind(
        group: String,
        itemClickListener: (String) -> Unit,
        addListener: (String) -> Unit,
        deleteListener: (String, Int) -> Unit,
    ) {
        binderHelper.bind(binding.swipeRevealLayout, group)
        binding.tvGroup.text = group
        binding.frontLayout.setOnClickListener { itemClickListener.invoke(group) }
        binding.add.setOnClickListener {
            addListener.invoke(group)
            binding.swipeRevealLayout.close(true)
        }
        binding.delete.setOnClickListener {
            deleteListener.invoke(group, adapterPosition)
            binding.swipeRevealLayout.close(true)
        }
    }
}
